docs: add flow of funds docs (#1644)
interledger/rafiki: packages/documentation/docs/integration/flow-of-funds.md

> # Flow of funds

Reviewer: ```suggestion
# Flow of Funds
```

> 

Reviewer: Should there be an intro (or a mermaid diagram) here of a typical payment flow, ie, 
1) create incoming payment on receiving ASE
2) create quote on sending ASE
3) create outgoing payment on sending ASE

?

We don't really have a section for this explanation

Author: Does this still apply given that it's now an event handler doc and not a doc on the flow of funds?

Reviewer 2: I think the current state of the Event Handlers section covers most of what's necessary - what actions the ASE should make in different situations.

Should we create a distinct section named `Communicating with another Rafiki instance` to address what Max mentioned, or would that overlap with the current OP client documentation?

> Rafiki communicates with the Account Servicing Entity over webhook events, which are defined in the _Getting Started_ section. Each event requires the Account Servicing Entity to either deposit or withdraw liquidity into or from Rafiki. This document will provide an example flow for each of the defines webhook events.

Reviewer 2: ```suggestion
Rafiki communicates with the Account Servicing Entity over webhook events, which are defined in the _Getting Started_ section. Each event requires the Account Servicing Entity to either deposit or withdraw liquidity into or from Rafiki. This document will provide an example flow for each of the defined webhook events.
```

> Rafiki communicates with the Account Servicing Entity over webhook events, which are defined in the _Getting Started_ section. Each event requires the Account Servicing Entity to either deposit or withdraw liquidity into or from Rafiki. This document will provide an example flow for each of the defines webhook events.

Reviewer: ```suggestion
Rafiki communicates with the Account Servicing Entity over webhook events, which are defined in the _Getting Started_ section. Most events require the Account Servicing Entity to either deposit or withdraw liquidity into or from Rafiki. This document will provide an example flow for each of the defines webhook events.
```
Not all events: `incoming_payment.created` (and by extension of events, the `payment_pointer.not_found` one

> Rafiki communicates with the Account Servicing Entity over webhook events, which are defined in the _Getting Started_ section. Each event requires the Account Servicing Entity to either deposit or withdraw liquidity into or from Rafiki. This document will provide an example flow for each of the defines webhook events.

Reviewer: ```suggestion
Rafiki communicates with the Account Servicing Entity over webhook events, which are defined in the [_Getting Started_](./getting-started) section. Each event requires the Account Servicing Entity to either deposit or withdraw liquidity into or from Rafiki. This document will provide an example flow for each of the defines webhook events.
```

## `incoming_payment.created`

> The `incoming_payment.created` event indicates that an incoming payment has been created. It has not yet received any funds so no action is required.

Reviewer: Should we mention here that (for example) you could notify a user of an upcoming incoming payment?

```mermaid
sequenceDiagram

participant ASE as Account Servicing Entity
participant R as Rafiki

R->>ASE: webhook event: incoming payment created
ASE->>ASE: no action required
```

## `incoming_payment.completed`

The `incoming_payment.completed` event indicates that an incoming payment has been completed, either automatically or manually, and that any funds that have been received into this incoming payment should be withdrawn and credited to the recipient's account with the Account Servicing Entity.

Example: An incoming payment was completed and received $10.

```mermaid
sequenceDiagram

participant ASE as Account Servicing Entity
participant R as Rafiki

> R->>ASE: webhook event: incoming payment completed,<br>receiveAmount: $10

Reviewer: ```suggestion
R->>ASE: webhook event: incoming payment completed,<br>receivedAmount: $10
```

ASE->>R: admin API call: WithdrawEventLiquidity
ASE->>ASE: credit receiver's account with $10
```

## `incoming_payment.expired`

The `incoming_payment.expired` event indicates that an incoming payment has expired, and that any funds that have been received into this incoming payment should be withdrawn and credited to the recipient's account with the Account Servicing Entity.

Example: An incoming payment has expired and received $2.55.

```mermaid
sequenceDiagram

participant ASE as Account Servicing Entity
participant R as Rafiki

> R->>ASE: webhook event: incoming payment expired,<br>receiveAmount: $2.55

Reviewer: ```suggestion
R->>ASE: webhook event: incoming payment expired,<br>receivedAmount: $2.55
```

ASE->>R: admin API call: WithdrawEventLiquidity
ASE->>ASE: credit receiver's account with $2.55
```

## `outgoing_payment.created`

The `outgoing_payment.created` event indicates that an outgoing payment has been created and is awaiting liquidity. The Account Servicing Entity should put a hold on the sender's account and deposit the funds into Rafiki.

Example: An outgoing payment for $12 has been created.

```mermaid
sequenceDiagram

participant ASE as Account Servicing Entity

> participant R as Rafiki

Reviewer: does it make sense to have this as "sending Rafiki"?

Author: Should we distinguish between sending and receiving? It may be confusing and people could think that there are two types of Rafiki

Reviewer 2: I don't think labeling this as "sending Rafiki" is necessary  since it's simply a part of the integration process. Currently, the ASE will only receive webhook events from one Rafiki instance, which is the one they intend to integrate.

R->>ASE: webhook event: outgoing payment created,<br>sendAmount: $12
ASE->>ASE: put hold of $12 on sender's account
ASE->>R: admin API call: DepositEventLiquidity
```

## `outgoing_payment.completed`

> The `outgoing_payment.completed` event indicates that an outgoing payment has successfully sent as many funds as possible to the receiver. The Account Servicing Entity should withdraw any access liquidity from that outgoing payment in Rafiki and book it as they see fit. One option would be to return it to the sender. The more likely scenario is that the access liquidity is considered as a fee and retained by the Account Servicing Entity. Furthermore, the Account Servicing Entity should remove the hold on the sender's account and debit it.

Reviewer: ```suggestion
The `outgoing_payment.completed` event indicates that an outgoing payment has successfully sent as many funds as possible to the receiver. The Account Servicing Entity should withdraw any access liquidity from that outgoing payment in Rafiki and book it as they see fit. One option would be to return it to the sender. The more likely scenario is that the access liquidity is considered as a fee and retained by the Account Servicing Entity. Furthermore, the Account Servicing Entity should remove the hold on the sender's account and debit it.
```
```suggestion
The `outgoing_payment.completed` event indicates that an outgoing payment has successfully sent as many funds as possible to the receiver. The Account Servicing Entity should withdraw any access liquidity from that outgoing payment in Rafiki and use it as they see fit. One option would be to return it to the sender. The more likely scenario is that the access liquidity is considered as a fee and retained by the Account Servicing Entity. Furthermore, the Account Servicing Entity should remove the hold on the sender's account and debit it.
```
Seems more straightforward?

> Example: An outgoing payment for $12 has completed. $11.50 were sent. The Account Servicing Entity keeps $0.50 as fees.

Reviewer 2: ```suggestion
Example: An outgoing payment for $12 has been completed. $11.50 were sent. The Account Servicing Entity keeps $0.50 as fees.
```

```mermaid
sequenceDiagram

participant ASE as Account Servicing Entity
participant R as Rafiki

R->>ASE: webhook event: outgoing completed,<br>sendAmount: $12, sentAmount:$11.50
ASE->>R: admin API call: WithdrawEventLiquidity

> ASE->>ASE: remove hold and deduct $12 from sender's account,<br>credit ASE's account with $0.50

Reviewer 2: ```suggestion
ASE->>ASE: remove the hold and deduct $12 from the sender's account,<br>credit ASE's account with $0.50
```

```

## `outgoing_payment.failed`

> The `outgoing_payment.failed` event indicates that an outgoing payment has either partially or completely failed and a retry was also unsuccessful. The Account Servicing Entity should withdraw any remaining liquidity from that outgoing payment in Rafiki. If the payment failed completely, the Account Servicing Entity should remove the hold from the sender's account. If the payment failed partially, the Account Servicing Entity should remove the hold from the sender's account and debit it with the amount that has been sent, but they should refrain from taking a sending fee.

Reviewer: ```suggestion
The `outgoing_payment.failed` event indicates that an outgoing payment has either partially or completely failed and a retry was also unsuccessful. The Account Servicing Entity should withdraw any remaining liquidity from that outgoing payment in Rafiki. If the payment failed completely (the sentAmount was 0), the Account Servicing Entity should remove the hold from the sender's account. If the payment failed partially, the Account Servicing Entity should remove the hold from the sender's account and debit it with the amount that has been sent, but they should refrain from taking a sending fee.
```
Suggestion

Example: An outgoing payment for $12 has failed. $8 were sent.

```mermaid
sequenceDiagram

participant ASE as Account Servicing Entity
participant R as Rafiki

R->>ASE: webhook event: outgoing failed,<br>sendAmount: $12, sentAmount:$8
ASE->>R: admin API call: WithdrawEventLiquidity

> ASE->>ASE: remove hold and deduct $8 from sender's account

Reviewer 2: ```suggestion
ASE->>ASE: remove the hold and deduct $8 from the sender's account
```
